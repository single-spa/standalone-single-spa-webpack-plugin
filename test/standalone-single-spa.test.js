const StandaloneSingleSpa = require("../lib/standalone-single-spa");
const HtmlWebpackPlugin = require("html-webpack-plugin");
const webpack = require("webpack");
const path = require("path");
const StandalonePlugin = require("../lib/standalone-single-spa");
const fs = require("fs").promises;

describe("standalone-single-spa-webpack-plugin", () => {
  test("basic-usage", async () => {
    const outputDir = path.resolve(__dirname, "./output/basic-usage");

    const config = {
      entry: path.resolve(__dirname, "./fixtures/basic/index.js"),
      output: {
        libraryTarget: "system",
        path: outputDir,
      },
      plugins: [
        new HtmlWebpackPlugin(),
        new StandalonePlugin({
          appOrParcelName: "basic-usage",
          importMapUrl: new URL(
            "https://react.microfrontends.app/importmap.json"
          ),
          importMap: {
            imports: {
              foo: "/foo.js",
            },
          },
        }),
      ],
    };

    const stats = await webpackCompile(config);
    const html = await readOutputHtml(outputDir);
    expect(html).toMatchSnapshot();
  });

  test("basic-parcel", async () => {
    const outputDir = path.resolve(__dirname, "./output/basic-parcel");

    const config = {
      entry: path.resolve(__dirname, "./fixtures/basic/index.js"),
      output: {
        libraryTarget: "system",
        path: outputDir,
      },
      plugins: [
        new HtmlWebpackPlugin(),
        new StandalonePlugin({
          isParcel: true,
          appOrParcelName: "basic-parcel",
          importMapUrl: new URL(
            "https://react.microfrontends.app/importmap.json"
          ),
          importMap: {
            imports: {
              foo: "/foo.js",
            },
          },
        }),
      ],
    };

    const stats = await webpackCompile(config);
    const html = await readOutputHtml(outputDir);
    expect(html).toMatchSnapshot();
  });

  test("disabled", async () => {
    const outputDir = path.resolve(__dirname, "./output/basic-parcel");

    const config = {
      entry: path.resolve(__dirname, "./fixtures/basic/index.js"),
      output: {
        libraryTarget: "system",
        path: outputDir,
      },
      plugins: [
        new HtmlWebpackPlugin(),
        new StandalonePlugin({
          isParcel: true,
          appOrParcelName: "basic-parcel",
          importMapUrl: new URL(
            "https://react.microfrontends.app/importmap.json"
          ),
          importMap: {
            imports: {
              foo: "/foo.js",
            },
          },
          disabled: true,
        }),
      ],
    };

    const stats = await webpackCompile(config);
    const html = await readOutputHtml(outputDir);
    expect(html.includes("systemjs")).toBe(false);
  });

  test("disabled import-map-overrides", async () => {
    const outputDir = path.resolve(__dirname, "./output/basic-usage");

    const config = {
      entry: path.resolve(__dirname, "./fixtures/basic/index.js"),
      output: {
        libraryTarget: "system",
        path: outputDir,
      },
      plugins: [
        new HtmlWebpackPlugin(),
        new StandalonePlugin({
          appOrParcelName: "basic-usage",
          importMapOverrides: false,
        }),
      ],
    };

    const stats = await webpackCompile(config);
    const html = await readOutputHtml(outputDir);
    expect(html).toMatchSnapshot();
  });

  test("import-map-overrides with local storage key", async () => {
    const outputDir = path.resolve(__dirname, "./output/basic-usage");

    const config = {
      entry: path.resolve(__dirname, "./fixtures/basic/index.js"),
      output: {
        libraryTarget: "system",
        path: outputDir,
      },
      plugins: [
        new HtmlWebpackPlugin(),
        new StandalonePlugin({
          appOrParcelName: "basic-usage",
        }),
      ],
    };

    const stats = await webpackCompile(config);
    const html = await readOutputHtml(outputDir);
    expect(html).toMatchSnapshot();
  });

  test("customProps option", async () => {
    const outputDir = path.resolve(__dirname, "./output/custom-props");

    const config = {
      entry: path.resolve(__dirname, "./fixtures/basic/index.js"),
      output: {
        libraryTarget: "system",
        path: outputDir,
      },
      plugins: [
        new HtmlWebpackPlugin(),
        new StandalonePlugin({
          appOrParcelName: "basic-usage",
          importMapOverridesLocalStorageKey: "devtools",
          customProps: {
            hi: "there",
          },
        }),
      ],
    };

    const stats = await webpackCompile(config);
    const html = await readOutputHtml(outputDir);
    expect(html).toMatchSnapshot();
  });

  test("startOptions option", async () => {
    const outputDir = path.resolve(__dirname, "./output/custom-props");

    const config = {
      entry: path.resolve(__dirname, "./fixtures/basic/index.js"),
      output: {
        libraryTarget: "system",
        path: outputDir,
      },
      plugins: [
        new HtmlWebpackPlugin(),
        new StandalonePlugin({
          appOrParcelName: "basic-usage",
          startOptions: {
            urlRerouteOnly: false,
          },
        }),
      ],
    };

    const stats = await webpackCompile(config);
    const html = await readOutputHtml(outputDir);
    expect(html).toMatchSnapshot();
  });

  test("multiple import map urls", async () => {
    const outputDir = path.resolve(__dirname, "./output/basic-usage");

    const config = {
      entry: path.resolve(__dirname, "./fixtures/basic/index.js"),
      output: {
        libraryTarget: "system",
        path: outputDir,
      },
      plugins: [
        new HtmlWebpackPlugin(),
        new StandalonePlugin({
          appOrParcelName: "basic-usage",
          importMapUrls: [
            new URL("https://react.microfrontends.app/importmap.json"),
            new URL("https://vue.microfrontends.app/importmap.json"),
          ],
        }),
      ],
    };

    const stats = await webpackCompile(config);
    const html = await readOutputHtml(outputDir);
    expect(html).toMatchSnapshot();
  });

  test("public path", async () => {
    const outputDir = path.resolve(__dirname, "./output/basic-usage");

    const config = {
      entry: path.resolve(__dirname, "./fixtures/basic/index.js"),
      output: {
        libraryTarget: "system",
        path: outputDir,
        publicPath: "/testpublicpath/",
      },
      plugins: [
        new HtmlWebpackPlugin(),
        new StandalonePlugin({
          appOrParcelName: "basic-usage",
          importMapUrls: [
            new URL("https://react.microfrontends.app/importmap.json"),
            new URL("https://vue.microfrontends.app/importmap.json"),
          ],
        }),
      ],
    };

    const stats = await webpackCompile(config);
    const html = await readOutputHtml(outputDir);
    expect(html).toMatchSnapshot();
  });
});

function webpackCompile(config) {
  return new Promise((resolve, reject) => {
    webpack(config, (err, stats) => {
      if (err) {
        reject(err);
      } else if (stats.compilation.errors.length > 0) {
        reject(stats.compilation.errors);
      } else {
        resolve(stats);
      }
    });
  });
}

async function readOutputHtml(outputDir) {
  return await fs.readFile(path.resolve(outputDir, "./index.html"), "utf-8");
}
