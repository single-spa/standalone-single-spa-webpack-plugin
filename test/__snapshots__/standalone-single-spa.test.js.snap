// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`standalone-single-spa-webpack-plugin basic-parcel 1`] = `
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>
      Webpack App
    </title>
    <meta name="viewport"
          content="width=device-width,initial-scale=1"
    >
  </head>
  <body>
    <script type="systemjs-importmap"
            src="https://react.microfrontends.app/importmap.json"
    >
    </script>
    <script type="systemjs-importmap">
      {
  "imports": {
    "single-spa": "https://cdn.jsdelivr.net/npm/single-spa/lib/system/single-spa.dev.js",
    "basic-parcel": "/main.js",
    "foo": "/foo.js"
  }
}
    </script>
    <script src="https://cdn.jsdelivr.net/npm/regenerator-runtime@0.13.7/runtime.min.js">
    </script>
    <script src="https://cdn.jsdelivr.net/npm/systemjs/dist/system.js">
    </script>
    <script src="https://cdn.jsdelivr.net/npm/systemjs/dist/extras/amd.js">
    </script>
    <script>
      Promise.all([System.import('single-spa'), System.import('basic-parcel')]).then(function (values) {
        const singleSpa = values[0];
        const parcelConfig = values[1];
        const parcelContainer = document.createElement('div');
        parcelContainer.id = 'parcel-container';
        document.body.appendChild(parcelContainer);
        singleSpa.mountRootParcel(parcelConfig, { domElement: parcelContainer, name: 'basic-parcel' });
      });
    </script>
  </body>
</html>
`;

exports[`standalone-single-spa-webpack-plugin basic-usage 1`] = `
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>
      Webpack App
    </title>
    <meta name="viewport"
          content="width=device-width,initial-scale=1"
    >
  </head>
  <body>
    <script type="systemjs-importmap"
            src="https://react.microfrontends.app/importmap.json"
    >
    </script>
    <script type="systemjs-importmap">
      {
  "imports": {
    "single-spa": "https://cdn.jsdelivr.net/npm/single-spa/lib/system/single-spa.dev.js",
    "basic-usage": "/main.js",
    "foo": "/foo.js"
  }
}
    </script>
    <script src="https://cdn.jsdelivr.net/npm/regenerator-runtime@0.13.7/runtime.min.js">
    </script>
    <script src="https://cdn.jsdelivr.net/npm/systemjs/dist/system.js">
    </script>
    <script src="https://cdn.jsdelivr.net/npm/systemjs/dist/extras/amd.js">
    </script>
    <script>
      System.import('single-spa').then(function (singleSpa) {
  singleSpa.registerApplication({
    name: 'basic-usage',
    app: function () {
      return System.import('basic-usage');
    },
    activeWhen: ["/"]
  });
  if (!window.location.pathname.startsWith('/')) {
    singleSpa.navigateToUrl('/');
  }
  singleSpa.start();
});
    </script>
  </body>
</html>
`;
